docs/source/python_tutorials/fsa_algo.rst

```diff
@@ -0,0 +1,217 @@
+Fsa algorithms
+==============
+
+This tutorial describes algorithms supported by k2.
+Most of the algorithms support both CPU and CUDA.
+A few of them support only CPU, which is documented
+explicitly in the corresponding documentation.
+
+.. NOTE::
+
+  All algorithms support FsaVec. A few of them also
+  support single FSAs.
+
+invert
+~~~~~~
+
+:func:`k2.invert` swaps the ``labels`` and ``aux_labels`` of a :class:`k2.Fsa`.
+
+The following is an example swapping the ``labels`` and ``aux_labels`` of an FSA.
+
+.. literalinclude:: ./code/invert1.py
+  :caption: Example code for :func:`k2.invert`
+  :language: python
+  :lines: 2-
+
+Its outputs are given below:
+
+.. figure:: images/before_invert.svg
+    :alt: before_invert
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/after_invert.svg
+    :alt: after_invert
+    :align: center
+    :figwidth: 600px
+
+
+When ``aux_labels`` is a ragged tensor, new arcs will be added after the invert.
+
+The following code applies :func:`k2.invert` to an FSA
+with ragged tensors as ``aux_labels``.
+
+.. literalinclude:: ./code/invert2.py
+  :caption: :func:`k2.invert` for FSAs with ragged tensors as ``aux_labels``
+  :language: python
+  :lines: 2-
+
+.. figure:: images/before_invert_aux.svg
+    :alt: before_invert_aux
+    :align: center
+    :figwidth: 600px
+
+.. figure:: images/after_invert_aux.svg
+    :alt: after_invert_aux
+    :align: center
+    :figwidth: 600px
+
+    Example of :func:`k2.invert` for FSAs with ragged tensors as ``aux_labels``.
+
+.. NOTE::
+
+  - :func:`k2.invert` supports CUDA as well as CPU.
+  - autograd is also supported.
+  - Its input can be either a single FSA or a FsaVec.
+
+
+arc_sort
+~~~~~~~~
+
+:func:`k2.intersect` and :func:`k2.compose` require two
+arc-sorted inputs. You can use :func:`k2.arc_sort` to convert
+an unsorted FSA/FsaVec into a sorted FSA/FsaVec.
+
+An FSA is sorted if for each state, its leaving arcs are sorted
```

Comment:
In k2 now, what we really do in arc-sort is: sort on label first then on dest_state